### PR TITLE
LOK-2478 : Syslog message from camel route to Syslog SinkModule

### PR DESCRIPTION
**Pull request by MusaidAli**
## Description
Syslog message from camel route to Syslog SinkModule

## Jira link(s)
- https://opennms.atlassian.net/browse/LOK-2478

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: minion-gateway/main/src/main/java/org/opennms/miniongateway/grpc/server/syslog/SyslogKafkaForwarder.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    public static final String DEFAULT_TRAP_RESULTS_TOPIC = "traps";
+
+    private final SinkMessageKafkaPublisher<SyslogMessageLogDTO, TenantLocationSpecificSyslogLogDTO> kafkaPublisher;
+
+    @Autowired
+    public SyslogKafkaForwarder(
+            SinkMessageKafkaPublisherFactory messagePublisherFactory,
+            TenantLocationSpecificSyslogMessageDTOMapper mapper,
+            @Value("${traps.results.kafka-topic:" + DEFAULT_TRAP_RESULTS_TOPIC + "}") String kafkaTopic) {
```
Comment (fooker) [inline minion-gateway/main/src/main/java/org/opennms/miniongateway/grpc/server/syslog/SyslogKafkaForwarder.java:49]: Same here. This ain't not traps

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java [hunk 2]
```diff
@@ -77,6 +108,46 @@ public boolean process(Exchange exchange, AsyncCallback callback) {
 
                         LOG.info(source.getHostName());
                         LOG.info("Message received : " + exchange.getIn().getBody());
+
+                        try {
+                            String content = buffer.toString(StandardCharsets.UTF_8);
+                            LOG.info("content" + content);
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java:114]: Please reduce log level and amount of messages

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java [hunk 3]
```diff
@@ -77,6 +108,46 @@ public boolean process(Exchange exchange, AsyncCallback callback) {
...
+                            SyslogMessage m_message = syslogParser.parse(content);
+                            LOG.info("Severity:" + m_message.getSeverity());
+
+                            LOG.info("Message:" + m_message.getMessage());
+                            LOG.info("Facility:" + m_message.getFacility());
+                            LOG.info("Host:" + m_message.getHostname());
+                            LOG.info("Date:" + m_message.getDate().toString());
+
+                            SyslogIdentity syslogIdentity = SyslogIdentity.newBuilder()
+                                    .setEnterpriseId("EntID")
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java:124]: Still no enterprise ID for syslog.

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java [hunk 4]
```diff
@@ -77,6 +108,46 @@ public boolean process(Exchange exchange, AsyncCallback callback) {
 
                         LOG.info(source.getHostName());
                         LOG.info("Message received : " + exchange.getIn().getBody());
+
+                        try {
+                            String content = buffer.toString(StandardCharsets.UTF_8);
+                            LOG.info("content" + content);
+                            SyslogMessage m_message = syslogParser.parse(content);
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java:115]: Variable naming: `m_` in members is a pity, but as a variable it's a no-no.

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java [hunk 5]
```diff
@@ -77,6 +108,46 @@ public boolean process(Exchange exchange, AsyncCallback callback) {
...
+                            LOG.info("Message:" + m_message.getMessage());
+                            LOG.info("Facility:" + m_message.getFacility());
+                            LOG.info("Host:" + m_message.getHostname());
+                            LOG.info("Date:" + m_message.getDate().toString());
+
+                            SyslogIdentity syslogIdentity = SyslogIdentity.newBuilder()
+                                    .setEnterpriseId("EntID")
+                                    .setGeneric(1)
+                                    .build();
+                            Identity identity1 = Identity.newBuilder()
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java:127]: What is this identity and where is `identity2`?

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java [hunk 6]
```diff
@@ -77,6 +108,46 @@ public boolean process(Exchange exchange, AsyncCallback callback) {
...
+                                    .setSystemId("SyslogID")
+                                    .build();
+                            SyslogMessageDTO message = SyslogMessageDTO.newBuilder()
+                                    .setRawMessage(ByteString.copyFrom(
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java:131]: You go the whole loop from bytes to string to parsed message to string to bytes for nothing

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java [hunk 7]
```diff
@@ -77,6 +108,46 @@ public boolean process(Exchange exchange, AsyncCallback callback) {
...
+
+                            SyslogMessageLogDTO dto = SyslogMessageLogDTO.newBuilder()
+                                    .setSyslogAddress(source.getHostName())
+                                    .setIdentity(identity1)
+                                    .setSyslogAddress(source.getHostName())
+                                    .addSyslogMessageDTO(message)
+                                    .build();
+                            getMessageDispatcher().send(dto);
+
+                            LOG.info("Message sent");
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java:144]: Remove useless log messages

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java [hunk 8]
```diff
@@ -77,6 +108,46 @@ public boolean process(Exchange exchange, AsyncCallback callback) {
...
+
+                        } catch (final Exception ex) {
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java:146]: Again, do not catch raw exception!

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/parser/SyslogMessage.java [hunk 9]
```diff
@@ -30,11 +31,23 @@
...
+
+    private Integer year;
+    private Integer month;
+    private Integer dayOfMonth;
+    private Integer hourOfDay;
+    private Integer minute;
+    private Integer second;
+    private Integer millisecond;
+    // private ZoneId zoneId;
+    // will add
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/parser/SyslogMessage.java:47]: Why date and separate fields?

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/parser/SyslogParser.java [hunk 10]
```diff
@@ -21,37 +21,207 @@
...
 public class SyslogParser {
     private static final Logger LOG = LoggerFactory.getLogger(SyslogParser.class);
     private static final Pattern m_messageIdPattern = Pattern.compile("^((\\S+):\\s*)");
-    // date pattern has been updated to support space at start and end of
-    // the message due to which the date match failed and current system date
-    // used to be passed which cause 3ms more time in parsing message
     private static final Pattern m_datePattern = Pattern.compile("^\\s*((\\d\\d\\d\\d-\\d\\d-\\d\\d)\\s*)");
     private static final Pattern m_oldDatePattern =
             Pattern.compile("^\\s*(\\S\\S\\S\\s+\\d{1,2}\\s+\\d\\d:\\d\\d:\\d\\d)\\s+");
+    private static final String datePattern = "((19|20)\\d{2})-([1-9]|0[1-9]|1[0-2])-(0[1-9]|[1-9]|[12][0-9]|3[01])";
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/parser/SyslogParser.java:45]: Does this replace `m_datePattern`. What about the comment above - why drop the performance improvement?

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/parser/SyslogParser.java [hunk 11]
```diff
@@ -21,37 +21,207 @@
...
+                LOG.trace("Found message '{}'", matchedMessage);
+            }
+
+            syslogMessage.setHostname(m.group(m_matchingGroupHost));
+
+            message = matchedMessage;
+        } else {
+            LOG.info("Regexp not matched: {}", message);
+            return null;
+        }*/
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/parser/SyslogParser.java:153]: What about this code?

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/parser/SyslogParser.java [hunk 12]
```diff
@@ -21,37 +21,207 @@
...
         // Extracting header and message parts
         String[] parts = message.substring(endIndex).split(" ", 6);
 
         // Assigning parts to fields (simplified parsing, assuming well-formed messages)
-        entry.setVersion(parts[0]);
-        entry.setTimestamp(parts[1]);
-        entry.setHostname(parts[2]);
-        entry.setApplication(parts[3]);
-        entry.setMessageId(parts[4]);
-        entry.setMessage(parts[5]);
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/parser/SyslogParser.java:54]: THB, I don't get why this change is necessary. Is this ported from classic?

File: minion/minion-syslog/listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 13]
```diff
@@ -1,28 +1,60 @@
...
+        <cm:default-properties>
+            <cm:property name="syslog.listen.interface" value="0.0.0.0" />
+            <cm:property name="syslog.listen.port" value="20514" />
+            <cm:property name="syslog.threads" value="10" />
+            <cm:property name="syslog.queue.size" value="10000" />
+            <cm:property name="syslog.batch.size" value="1000" />
+            <cm:property name="syslog.batch.interval" value="500" />
+            <cm:property name="syslog.m_forwardingPattern" value="^.*\s*(19|20)\d\d([-/.])(0[1-9]|1[012])\2(0[1-9]|[12][0-9]|3[01])(\s+)(\S+)(\s)(\S.+)" />
+            <cm:property name="syslog.m_matchingGroupHost" value="6" />
+            <cm:property name="syslog.m_matchingGroupMessage" value="8" />
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml:26]: Please remove `m_`

File: minion/minion-syslog/listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 14]
```diff
@@ -1,28 +1,60 @@
...
+        <property name="numThreads" value="${syslog.threads}" />
+        <property name="queueSize" value="${syslog.queue.size}" />
+        <property name="batchSize" value="${syslog.batch.size}" />
+        <property name="batchIntervalMs" value="${syslog.batch.interval}" />
+    </bean>
 
+    <bean id="syslogParser" class="org.opennms.horizon.minion.syslog.listener.parser.SyslogParser">
+        <property name="m_forwardingPattern" value="${syslog.m_forwardingPattern}" />
+        <property name="m_matchingGroupHost" value="${syslog.m_matchingGroupHost}" />
+        <property name="m_matchingGroupMessage" value="${syslog.m_matchingGroupMessage}" />
```
Comment (fooker) [inline minion/minion-syslog/listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml:42]: Please remove `m_`

File: shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+import "snmp-api.proto";
+import "opennms_minion_ipc.proto";
+
+
+message SyslogIdentity {
+  int32 generic = 1;
+  int32 specific = 2;
+  string enterpriseId = 3;
+  string syslogOID = 4;
```
Comment (fooker) [inline shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto:16]: What is all this? AFAIK there is no enterpriseId of OID in syslog. And what does generic and specific even mean?

File: shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+message SyslogMessageDTO {
+  string syslogAddress = 1;
+  string port = 2 ;
+  string systemId = 3;
+  string location = 4 ;
+  bytes rawMessage = 5;
+  string community = 6;
```
Comment (fooker) [inline shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto:25]: No community in syslog.

File: shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  string version = 7;
```
Comment (fooker) [inline shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto:26]: What version is this?

File: shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/tenantLocationSyslog.proto [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+
+import "opennms/syslog/syslog.proto";
+import "opennms_minion_ipc.proto";
+
+message TenantLocationSpecificSyslogLogDTO {
+  string tenant_id = 100;
+  string location_id = 101;
+
+  minion.Identity identity = 1;
+  string syslogAddress = 2;
```
Comment (fooker) [inline shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/tenantLocationSyslog.proto:16]: What is `syslogAddress`? Reveiver address? sender address?

File: minion-gateway/main/src/main/java/org/opennms/miniongateway/grpc/server/syslog/SyslogKafkaForwarder.java [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+/**
+ * Forwarder of Syslog messages - received via GRPC and forwarded to Kafka.
+ */
+@Component
+public class SyslogKafkaForwarder implements MessageConsumer<SyslogMessageLogDTO, SyslogMessageLogDTO> {
+
+    public static final String DEFAULT_TRAP_RESULTS_TOPIC = "traps";
```
[the remark this answers is not in the document]
Comment (MusaidAli) [reply minion-gateway/main/src/main/java/org/opennms/miniongateway/grpc/server/syslog/SyslogKafkaForwarder.java]: added new topic syslog.please check latest commit.

File: minion-gateway/main/src/main/java/org/opennms/miniongateway/grpc/server/syslog/SyslogKafkaForwarder.java (hunk 1 above)
Comment (MusaidAli) [reply minion-gateway/main/src/main/java/org/opennms/miniongateway/grpc/server/syslog/SyslogKafkaForwarder.java]: resolved.

File: minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java (hunk 4 above)
Comment (MusaidAli) [reply minion/minion-syslog/listener/src/main/java/org/opennms/horizon/minion/syslog/listener/SyslogRoute.java:114]: resolved.

File: minion/minion-syslog/listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 14 above)
Comment (MusaidAli) [reply minion/minion-syslog/listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: resolved.

File: shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto (hunk 15 above)
Comment (MusaidAli) [reply shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto]: resolved.please check latest commit.

File: shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto (hunk 17 above)
Comment (MusaidAli) [reply shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto]: resolved.

File: shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/tenantLocationSyslog.proto (hunk 18 above)
Comment (MusaidAli) [reply shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/tenantLocationSyslog.proto]: resolved.

File: shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto (hunk 16 above)
Comment (MusaidAli) [reply shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto]: resolved.

File: shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto [hunk 20]
```diff
@@ -8,28 +8,17 @@ option java_package = "org.opennms.horizon.grpc.syslog.contract";
...
-  int32 specific = 2;
-  string enterpriseId = 3;
-  string syslogOID = 4;
-}
-
 message SyslogMessageDTO {
-  string syslogAddress = 1;
-  string port = 2 ;
+  string listenAddress = 1;
+  string listenPort = 2 ;
```
Comment (fooker) [inline shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/syslog.proto:13]: Sending back the listening address and port here sounds useless. Do you mean to send the sender infos?

File: shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/tenantLocationSyslog.proto [hunk 21]
```diff
@@ -13,6 +12,8 @@ message TenantLocationSpecificSyslogLogDTO {
   string location_id = 101;
 
   minion.Identity identity = 1;
-  string syslogAddress = 2;
+  string listenAddress = 2;
```
Comment (fooker) [inline shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/tenantLocationSyslog.proto:15]: If we have this info here and in the repeated inner message, which one do we use?

File: events/syslog/src/main/java/org/opennms/horizon/events/syslog/impl/EventXmlToProtoMapperImpl.java [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+import org.opennms.horizon.events.proto.EventParameter;
+import org.opennms.horizon.events.syslog.EventXmlToProtoMapper;
+import org.opennms.horizon.events.xml.Event;
+import org.opennms.horizon.events.xml.Parm;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+@Component
+public class EventXmlToProtoMapperImpl implements EventXmlToProtoMapper {
```
Comment (fooker) [inline events/syslog/src/main/java/org/opennms/horizon/events/syslog/impl/EventXmlToProtoMapperImpl.java:36]: We already have these somewhere. share them

**Review by fooker — CHANGES_REQUESTED** [review]
I don't see the point in duplicating the whole event sending code. Can you make the existing one shared?

File: shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/tenantLocationSyslog.proto (hunk 18 above)
Comment (MusaidAli) [reply shared-lib/lokahi-ipc/ipc-grpc/ipc-grpc-contract/src/main/proto/opennms/syslog/tenantLocationSyslog.proto]: added senderAddress

**Pull request closed without merging**